cli_tools/diagnostics/main_windows.go

```diff
@@ -0,0 +1,248 @@
+//  Copyright 2018 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+)
+
+type cmd struct {
+	path           string
+	args           string
+	outputFileName string
+	// True when the command produces its own file and doesn't need one
+	// created from stdout.
+	cmdProducesFile bool
+}
+
+type wmiQuery struct {
+	class          string
+	namespace      string
+	outputFileName string
+}
+
+func (command cmd) run() (string, error) {
+	outPath := filepath.Join(tmpFolder, command.outputFileName)
+
+	c := exec.Command(command.path)
+	argString := command.args
+
+	if command.cmdProducesFile {
+		// Replace any output file args with that path in a temp folder
+		relPath := command.outputFileName
+		argString = strings.Replace(argString, relPath, outPath, -1)
+	} else {
+		// If the command doesn't produce a file, we need to construct
+		// one from Stdout and Stderr
+		outFile, err := os.Create(outPath)
+		if err != nil {
+			log.Printf("Error creating file %s: %v", outPath, err)
+			return outPath, err
+		}
+		defer outFile.Close()
+		c.Stdout = outFile
+		c.Stderr = outFile
+	}
+
+	if command.args != "" {
+		c.Args = append(c.Args, strings.Split(argString, " ")...)
+	}
+	err := c.Run()
+	return outPath, err
+}
+
+func (query wmiQuery) run() (string, error) {
+	outPath := filepath.Join(tmpFolder, query.outputFileName)
+	outFile, err := os.Create(outPath)
+	if err != nil {
+		return outPath, err
+	}
+	defer outFile.Close()
+
+	// WMI is somewhat flaky, so we should retry a few times on failures
+	var data string
+	for i := 0; i < 3; i++ {
+		data, err = printWmiObjects(query.class, query.namespace)
+		if err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return outPath, err
+	}
+
+	header := fmt.Sprintf("Queried wmi objects [%s] from namespace %s\n\n", query.class, query.namespace)
+	if _, err = outFile.WriteString(header); err != nil {
+		return outPath, err
+	}
+
+	_, err = outFile.WriteString(data)
+	return outPath, err
+}
+
+func runAll(commands []runner, errCh chan error) []string {
+	paths := make([]string, 0, len(commands))
+	ch := make(chan string, len(commands))
+	errs := make(chan error, len(commands))
+	var wg sync.WaitGroup
+
+	for _, command := range commands {
```

Comment:
Your doing a lot of concurrent runs in this tool making it very resource intensive, maybe serialize runAll? I'm on the fence about serializing gatherLogs too.

Reply:
Changed runAll to be serial. Keeping gatherLogs in concurrent as some commands like tracing will take multiple minutes.